File: example/src/main/java/jp/bglb/bonboru/flux/example/ui/MainActivity.java
package jp.bglb.bonboru.flux.example.ui;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;
import jp.bglb.bonboru.flux.Dispatcher;
import jp.bglb.bonboru.flux.example.MyDispatcher;
import jp.bglb.bonboru.flux.example.R;
import jp.bglb.bonboru.flux.example.action.ActionTypes;
import jp.bglb.bonboru.flux.example.action.MainAction;
import jp.bglb.bonboru.flux.example.action.MainAction2;
import jp.bglb.bonboru.flux.example.action.MainAction3;
import jp.bglb.bonboru.flux.example.action.MainAction4;
import jp.bglb.bonboru.flux.example.dto.MainData;
import jp.bglb.bonboru.flux.example.dto.MainDataStore;
import jp.bglb.bonboru.flux.example.middleware.LoadMiddleware;
import jp.bglb.bonboru.flux.example.reducer.MainReducer;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action1;
import rx.schedulers.Schedulers;
import rx.subscriptions.CompositeSubscription;

/**
 * Created by tmasuda on 2016/04/15.
 */
public class MainActivity extends AppCompatActivity {

  MainDataStore store = new MainDataStore();

  MainReducer mainReducer = new MainReducer();

  MainAction action = new MainAction();

  MainAction2 action2 = new MainAction2();

  MainAction3 action3 = new MainAction3();

  MainAction4 action4 = new MainAction4();

  Dispatcher<MainData, ActionTypes> dispatcher;

  TextView message;

  Button button;
  Button button2;
  Button button3;
  ProgressBar progressBar;

  CompositeSubscription subscription;

  @Override protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);
    message = (TextView) findViewById(R.id.message);
    button = (Button) findViewById(R.id.button);
    button2 = (Button) findViewById(R.id.button2);
    button3 = (Button) findViewById(R.id.button3);
    progressBar = (ProgressBar) findViewById(R.id.progress);
    dispatcher = new MyDispatcher<>(mainReducer, store, new LoadMiddleware());
    button.setOnClickListener(new View.OnClickListener() {
      @Override public void onClick(View v) {
        dispatcher.dispatch(action4, action);
      }
    });
    button2.setOnClickListener(new View.OnClickListener() {
      @Override public void onClick(View v) {
        dispatcher.dispatch(action2);
      }
    });
    button3.setOnClickListener(new View.OnClickListener() {
      @Override public void onClick(View v) {
        dispatcher.dispatch(action3);
      }
    });
  }

  @Override protected void onResume() {
    super.onResume();
    subscription = new CompositeSubscription(store.message.subscribeOn(Schedulers.newThread())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(new Action1<String>() {
          @Override public void call(String str) {
            Log.d("Debug", "changed message");
            message.setText(str);
          }
        }), store.error.subscribeOn(Schedulers.newThread())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(new Action1<String>() {
          @Override public void call(String str) {
            Log.d("Debug", "changed error");
            if (str == null) {
              message.setVisibility(View.VISIBLE);
            } else {
              message.setVisibility(View.GONE);
              Toast.makeText(MainActivity.this, str, Toast.LENGTH_SHORT).show();
            }
          }
        }), store.progressBarVisibility.subscribeOn(Schedulers.newThread())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(new Action1<Boolean>() {
          @Override public void call(Boolean aBoolean) {
            progressBar.setVisibility(aBoolean ? View.VISIBLE : View.INVISIBLE);
          }
        }));
  }

  @Override protected void onPause() {
    super.onPause();
    subscription.unsubscribe();
  }
}

File: kotlinexample/src/main/kotlin/com/example/kotlinexample/ui/MainActivity.kt
package com.example.kotlinexample.ui

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import com.example.kotlinexample.R
import com.example.kotlinexample.action.MainActionCreator
import com.example.kotlinexample.action.MyActionType
import com.example.kotlinexample.data.MainData
import com.example.kotlinexample.data.MainDataStore
import com.example.kotlinexample.middleware.LoadMiddleware
import com.example.kotlinexample.reducer.MainReducer
import jp.bglb.bonboru.flux.kotlin.Dispatcher
import rx.android.schedulers.AndroidSchedulers
import rx.lang.kotlin.plusAssign
import rx.schedulers.Schedulers
import rx.subscriptions.CompositeSubscription

/**
 * Created by tetsuya on 2016/06/25.
 */
class MainActivity() : AppCompatActivity() {

  lateinit var message: TextView

  lateinit var button: Button

  lateinit var button2: Button

  lateinit var button3: Button

  lateinit var progressBar: ProgressBar

  lateinit var store: MainDataStore

  lateinit var reducer: MainReducer

  lateinit var dispatcher: Dispatcher<MainData, MyActionType>

  lateinit var subscription: CompositeSubscription

  var actionCreator = MainActionCreator()

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)
    message = findViewById(R.id.message) as TextView
    button = findViewById(R.id.button) as Button
    button2 = findViewById(R.id.button2) as Button
    button3 = findViewById(R.id.button3) as Button
    progressBar = findViewById(R.id.progress) as ProgressBar

    button.setOnClickListener { dispatcher.dispatch(actionCreator.action("hello")) }
    button2.setOnClickListener { dispatcher.dispatch(actionCreator.action("hello, Flux")) }
    button3.setOnClickListener { dispatcher.dispatch(actionCreator.action("hello, Flux, Kotlin")) }

    reducer = MainReducer()
    store = MainDataStore()
    dispatcher = Dispatcher(reducer, store, LoadMiddleware(), LoadMiddleware())
  }

  override fun onResume() {
    super.onResume()
    subscription = CompositeSubscription()
    subscription += store.mesage
        .subscribeOn(Schedulers.newThread())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe({
          message.text = it
        })

    subscription += store.progressVisibility
        .subscribeOn(Schedulers.newThread())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe({
          it?.let {
            progressBar.visibility = if (it) View.VISIBLE else View.INVISIBLE
          }
        })
  }

  override fun onPause() {
    super.onPause()
    subscription.unsubscribe()
  }
}
